### ISSUE TITLE
Proposed Design for #273 with Central Event Processor or ml-api-adapter

### ISSUE THREAD

**Issue by lewisdaly** [mojaloop/pisp#42]
I want to continue the discussion about [#273 - Transaction Notification Design](https://app.zenhub.com/workspaces/pisp-5e8457b05580fb04a7fd4878/issues/mojaloop/mojaloop/273). 

## Background & Assumptions

To reiterate some background knowledge and previous decisions:
1. We need some mechanism to inform the PISP that a Thirdparty Transaction has completed. Our decision is to use a pub/sub model managed by the switch to solve this problem
2. The switch will be _copying_ the `PUT /transfers/{id}` call from the _PayeeDFSP_ and _forwarding_ it to the _PISP_
3. **Subscription** is when a PISP is identified by the Switch as an _interested party_ in a future Transfer
4. **Notification** is when a Switch sends a `PUT /transfers/{id}` or `PUT /transfers/{id}/error` request to the _interested parties_
5. The mechanism for **Subscription** is as yet undefined, but we are thinking this could be during either the initial `POST /thirdpartyRequests/transactions` call, provided that the `transferId`  is set ahead of time by the PISP. Regardless of the decision here, we assume that some **Subscription** event will be generated by some service at some point in time before the transfer takes place.

## Options

The way I see it, there are 2 places we can put this implementation:
1. **Central Event Processor** - A service currently used for monitoring and alerts, but could also be used to generically watch events in Kafka, and emit other events
2. **ml-api-adapter notification handler** - The notification handler is already responsible for sending the `PUT /transfers/{ID}` to the PayerDFSP, so adding additional logic to send the **Notification** to the PISP could make sense here

*I think the Central Event Processor is the best approach, as I will detail below:*

### Option 1. `central-event-processor`

**Subscription Process**
1. `thirdparty-api-adapter` or some other service emits a `subscription` event on the `notification-topic`
1. `central-event-processor` consumes this message
1. `central-event-processor` saves a new **Subscription** object in it's database, with a `transferId` and `participantId`. `participantId` is the Id of the PISP to be informed when a transfer with the id of `transferId` is observed later

**Notification Process**
1. `central-ledger` emits a message on the `Notification-Topic`
  > Ref: [1.1.4.a-send-notification-to-participant](https://docs.mojaloop.io/documentation/mojaloop-technical-overview/central-ledger/transfers/1.1.4.a-send-notification-to-participant.html)
2. `central-event-processor` consumes the message
3. `central-event-processor` looks up the previous **Subscription** object, and retrieves the `participantId`
4. `central-event-processor` re-emits the message on the `Thirdparty-Transfer-Topic`
5. `central-event-processor` deletes the **Subscription** object (_this may not be necessary_)
6. `thirdparty-api-adapter` consumes message from `Thirdparty-Transfer-Topic`,
7. `thirdparty-api-adapter` sends the `PUT /transfers/...` to the PISP

**Implications:**
- Not in critical path, so it adds no overhead to transfer processing
- Takes advantage of our resilient messaging (kafka)
- It makes sense for the `central-event-processor` to  act as a simple _filter_ which filters incoming `PUT /transfers...` notifications and sends them through to the `thirdparty-api-adapter` if they are relevant to a previous `thirdpartyRequest
- This approach could be more complicated than #2, as there is more pieces and distributed messaging
- Potential duplication of business logic, which could make implementation over complicated and redundant
    - But it also means we can easily generalize for more than _just_ transfers with the `ml-api-adapter`
- Potentially slower than Option #2 - as it requires more internal messaging between services

### Option 2. `ml-api-adapter` notification handler

The existing notification handler could handle this logic, after sending the `PUT /transfers...` to the PayerDFSP

**Subscription Process**
1. `thirdparty-api-adapter` or some other service emits a `subscription` event on the `notification-topic`
1. `ml-api-adapter` consumes this message
1. `ml-api-adapter` saves a new **Subscription** object in it's database, with a `transferId` and `participantId`. `participantId` is the Id of the PISP to be informed when a transfer with the id of `transferId` is observed later

**Notification Process**
1. `central-ledger` emits a message on the `Notification-Topic`
  > Ref: [1.1.4.a-send-notification-to-participant](https://docs.mojaloop.io/documentation/mojaloop-technical-overview/central-ledger/transfers/1.1.4.a-send-notification-to-participant.html)
2. `ml-api-adapter` notification handler consumes this message, as normal
3. After step 14, it looks up some list of subscriptions to see if there are any PISPs interested in this transfer
4. Looks up the endpoint for the PISP, and sends the `PUT /transfers...` call

**Implications:**
- Also not in the critical path, so it would have no performance impact on transfers
- `ml-api-adapter` would need some sort of state to know about **Subscriptions**, and it currently is stateless
- Business logic is quite similar to the existing `PUT /transfers` calls, so it could be easy to implement in this case
    - But it doesn't generalize well for extensions to this approach, such as what is currently being proposed by the cross-network workstream
- This approach has the potential to be faster than Option 1, since there are less messages being exchanged internally between services

## Further Design on Option 1:

To get a feel for this problem a little bit more, and to understand how the Central Event Processor could be modified, I spent some time implementing a basic example of this request. You can see a working example [on my feature branch](https://github.com/vessels-tech/central-event-processor/tree/pisp/feature/273-tx-notification#pisp-thirdparty-notification-example)

The following sequence diagram captures both the Subscription and Notification Handling flow:

![](https://raw.githubusercontent.com/vessels-tech/central-event-processor/pisp/feature/273-tx-notification/docs/sequenceDiagrams/seq-cep-10.6-thirdparty-notifications-flow.svg)

### Ask

1. Given the above 2 options, **which approach do you think we should pursue?**
2. **Are there other ways you see we could tackle this problem?**
3. **If we choose option 1, does the Central Event Processor make sense? Or would it make sense to have a thirdparty-event-processor?** 
  > Note: My view on this would be to use the Central Event Processor, but make it easily configurable, and default to not listening/emitting 3rd party events.

**Comment by elnyry-sam-k** [comment]
Thanks Lewis, this is very comprehensive and captures the discussions about this quite nicely.

As we've discussed earlier, I'm leaning towards using the CEP (central-event-processor) as well; However, the subscriptions need not be managed at a message level I think, since this the association between FSPs and PISPs is not fleeting.. But some work needs to be done to make this part of the onboarding process for FSPs where they register the PISP(s) they're affiliated with (and registered to the Scheme) so that the rules in CEP are fairly long living.. 

I think in our discussion with Michael earlier, we almost decided to use not ml-api-adapter as its in the critical path and its best to not affect that at this stage. But, his is still a good discussion to have and arrive at a consensus..

**Comment by lewisdaly** [comment]
@elnyry-sam-k I do like the idea of having long-lived rules, but I'm not quite sure how it would work. 

The fundamental problem we are trying to solve here is we have a `PUT /transfers/{id}` message originating from a PayeeDFSP that knows _nothing_ about the PISP, and we need to forward that message to the PISP.

We could define rules based on some properties attached to the `PUT /transfers/{id}` message or headers, but that means coupling the ThirdpartyRequests logic with the PayeeDFSP, which I don't think we want to do.

Otherwise we might be able to rely on extension lists, but I don't think that's a great solution as then it's not 'baked in' to Mojaloop as it were.

Is there another way you could see how long lived subscriptions/rules could work?

**Comment by elnyry-sam-k** [comment]
Hi Lewis,

I still think we should be using something like `TransactionSubScenario` or `TransactionInitiatorType` in a `TransactionType` Object for a 'PISP' involved transaction (posting the definitions at the end for these two types along with `TransactionScenario` for reference). This way, when a quote happens, the PISP registers that transaction and then uses end-points that are already registered to the PISP based on headers and other information available and then uses it in the notifications (for transfers and may be quote responses even).

Definitions:
```
TransactionInitiator:
    title: TransactionInitiatorEnum
    type: string
    enum:
    - PAYER
    - PAYEE
    description: Below are the allowed values for the enumeration
      - PAYER Sender of funds is initiating the transaction. The account to send from is either owned by the Payer or is connected to the Payer in some way.
      - PAYEE Recipient of the funds is initiating the transaction by sending a transaction request. The Payer must approve the transaction, either automatically by a pre-generated OTP or by pre-approval of the Payee, or by manually approving in his or her own Device.

TransactionInitiatorType:
    title: TransactionInitiatorTypeEnum
    type: string
    enum:
    - CONSUMER
    - AGENT
    - BUSINESS
    - DEVICE
    description: Below are the allowed values for the enumeration
      - CONSUMER Consumer is the initiator of the transaction.
      - AGENT Agent is the initiator of the transaction.
      - BUSINESS Business is the initiator of the transaction.
      - DEVICE Device is the initiator of the transaction.

TransactionScenario:
    title: TransactionScenarioEnum
    type: string
    enum:
    - DEPOSIT
    - WITHDRAWAL
    - TRANSFER
    - PAYMENT
    - REFUND
    description: Below are the allowed values for the enumeration
      - DEPOSIT Used for performing a Cash-In (deposit) transaction. In a normal scenario, electronic funds are transferred from a Business account to a Consumer account, and physical cash is given from the Consumer to the Business User.
      - WITHDRAWAL Used for performing a Cash-Out (withdrawal) transaction. In a normal scenario, electronic funds are transferred from a Consumer’s account to a Business account, and physical cash is given from the Business User to the Consumer.
      - TRANSFER Used for performing a P2P (Peer to Peer, or Consumer to Consumer) transaction.
      - PAYMENT Usually used for performing a transaction from a Consumer to a Merchant or Organization, but could also be for a B2B (Business to Business) payment. The transaction could be online for a purchase in an Internet store, in a physical store where both the Consumer and Business User are present, a bill payment, a donation, and so on.
      - REFUND Used for performing a refund of transaction.

TransactionSubScenario:
    title: TransactionSubScenario
    type: string
    pattern: ^[A-Z_]{1,32}$
    description: Possible sub-scenario, defined locally within the scheme (UndefinedEnum Type).

TransactionType:
    title: TransactionType
    type: object
    description: Data model for the complex type TransactionType.
    properties:
      scenario:
        $ref: '#/definitions/TransactionScenario'
        description: Deposit, withdrawal, refund, …
      subScenario:
        $ref: '#/definitions/TransactionSubScenario'
        description: Possible sub-scenario, defined locally within the scheme.
      initiator:
        $ref: '#/definitions/TransactionInitiator'
        description: Who is initiating the transaction - Payer or Payee
      initiatorType:
        $ref: '#/definitions/TransactionInitiatorType'
        description: Consumer, agent, business, …
      refundInfo:
        $ref: '#/definitions/Refund'
        description: Extra information specific to a refund scenario. Should only be populated if scenario is REFUND
      balanceOfPayments:
        $ref: '#/definitions/BalanceOfPayments'
        description: Balance of Payments code.
    required:
    - scenario
    - initiator
    - initiatorType
```

**Comment by mjbrichards** [comment]
I had envisaged something like option 2. When the notification handler gets a message, it looks up the subscribers for that transaction. Then it sends the message to each of them. I agree, it needs the notification handler to look up the subscribers before deciding who gets the message, so it's no longer stateless. But it seems to me that this is a function which really does belong to the notification handler and not to the central event processor. I'd be against assigning it to a separate third-party processor, because that would make it less usable in other contexts. In particular, I envisaged something like the following scenario:

1. PISP sends a transactionRequest. This now includes a transaction ID (which the payer DFSP should use when it requests a quote...) Central event processor generates a subscription record for the sender of the message (the PISP) for this transaction.
2. Payer DFSP sends a quote request. Central event processor generates a subscription record for the sender for this transaction.
3. Payee DFSP sends a quote response. Notification handler looks up the subscribers to this transaction, who are, at this point, the Payer DFSP and the PISP. Both of them get the response (which means the payer DFSP doesn't have to include it in the authorisation call.)
4. Payer DFSP sends a transfer request.
5. Payee DFSP sends the response. If the status is RESERVED, then the central event processor generates a subcription record for the sender (that is, the payee DFSP) for this transaction. If not, not.
6. The notification handler sends the response to all currently subscribed parties: the PISP, the payer DFSP and the payee DFSP if it's requested it.

So the rule is that POST messages and GET messages are just passed on to the addressee; a PUT message is returned to all subscribers, or to the addressee if there are no subscribers (case for PUT /parties, for instance.) 

So that was my feeling about it: I wanted something that was relatively simple, and that could be used for all sorts of use cases. Let me know what you think.

**Comment by lewisdaly** [comment]
Thanks @MichaelJBRichards and @elnyry-sam-k .

To your points first, @MichaelJBRichards:

There are a couple benefits I thought of for using the Central Event Processor instead of using the notification topic itself:
1. (as you mentioned) This requires no changes to the notification handler. The additional callback behaviour is "bolted on" to the side, which I makes it more configurable and extendable for other use cases and callbacks in the future.
2. I feel like there should be an important distinction made between informing the DFSPs who are in the critical path of the transaction, and other 3rd parties who might want to receive PUT calls relevant to a transaction.
    - I do acknowledge however, that the optional `PATCH /transfers/{id}` to the PayeeDFSP does muddy the waters of this distinction however. I remember your original point was that the mechanism to send the `PATCH /transfers/{id}` should be the same as the `PUT /transfers/{id}` to the PISP
3. The notification handler lives inside the `ml-api-adapter`, so to me, it makes sense that the notification handler is only interested in converting messages related to the FSPIOP-API, and not the Thirdparty-API. I don't think it makes sense for the `ml-api-adapter` to know about Thirdparty related messags (or any other messages for other apis in the future).
    - The other way around this would be to have the subscription and callback logic for informing the PISPs in the thirdparty-api-adapter, but this would make our solution even harder to extend for other use cases in the future.

Regarding this point you made:
> I'd be against assigning it to a separate third-party processor, because that would make it less usable in other contexts.

Can you please expand on this a little more? I feel that a 3rd party processor makes it _more_ usable in other contexts, since no changes need to be made to the `ml-api-adapter` need to be made in order to listen in to messages within Mojaloop, and perform generic actions. Indeed this was the purpose of the Central Event Processor from the beginning with it's dynamic rules engines.


To summarize my feelings, I think the primary motivation for using the Central Event Processor is about separation of concerns between the different services:

- ml-api-adapter notification handler: 
    - Listens to `transaction-notification` events and issues callbacks to DFSPs
- thirdparty-api-adapter: 
    - listens for `thirdparty-notification` events, and issues callbacks to third parties
- Central Event Processor: 
    - listens for `notification-subscription` events and maintains some state about the open subscriptions
    - listens for `transaction-notification` events and _copies_ them to the `thirdparty-notification` topic if they match an open subscription

**Comment by lewisdaly** [comment]
@elnyry-sam-k if my memory serves correctly, we did talk about using the `TransactionSubScenario` or `TransactionInitiatorType` to distinguish between a PISP and non-PISP transaction, but I thought we decided against this because a PISP Transaction can fill (just about) all of those scenarios, and we didn't want to pidgeon hole the available transaction types for the PISP. 

Perhaps @MichaelJBRichards or @adrianhopebailie might be able to expand on this.

One of the other things I thought we wanted from this mechanism was to a generic way for third parties to subscribe to callbacks inside the switch, for example `PUT /consents/{id}` at the end of the account linking process. Such a stateless, rules based approach as you suggest could work for `PUT /quotes/{id}` and `PUT /transfers/{id}`, but it doesn't generalize as well.

**Comment by lewisdaly** [comment]
@jgeewax @MichaelJBRichards @elnyry-sam-k I'd very much love to pick up the conversation on this one - feel free to comment here, or I can schedule a design call next week if that's preferable.

**Comment by elnyry-sam-k** [comment]
Maybe a quick call to discuss would help, @lewisdaly

**Comment by eoln** [comment]
my, very very naive, 3 cents: 
- 1. why not use extensions structure - it is forwarded by the switch for every payload, am I right? then when PayerDFSP calls PUT /transfers/... for PISP scenario it adds proper extension (with information on how to reach PISP), then in notification handler, we look for the proper extension, and if found we notify PISP. So ml_api_adapter is still stateless
- 2. PayerDFSP already knows that it is PISP scenario, so why no to obligate it to notify all parties - the PayeeDFSP in one call and PISP in second -> so no changes in mojaloop core will be needed at all

**Comment by lewisdaly** [comment]
Thanks all for the discussion yesterday. 

One of the main things we focused on was the idea of _storing the list of recipients with the notification body_, instead of the above approach with the Pub/Sub model where the Central-Event-Processor receives subscription, and then waits around for an notification with a transactionId that matches something in it's subscription.

So perhaps we can redefine the options as follows:

1. Store the thirdparty subscriptions to a Notification separately to the Notification in and of itself. Have another service which listens for these notifications, and then broadcasts them to the thirdparty list
  > This is what we have with the CEP I outlined above

2. Store the thirdparty subscriptions to a Notification as a field in the notification itself. When the ml-api-adapter sees the notification, it sends it to it's primary recipient, then iterates through the list of thirdparty subscriptions, and sends the additional thirdparty notifications
  > The list of subscriptions here could be inferred from the original messages 

3. Make (optional) modifications to the API specification to explicitly state the thirdparty subscriptions parties up front. When the ml-api-adapter sees the notification, it sends it to it's primary recipient, then iterates through the list of thirdparty subscriptions, and sends the additional thirdparty notifications

**Comment by lewisdaly** [comment]
We made good progress last week on this, and are going ahead with the CEP approach.

Here's a PR with updated design: https://github.com/mojaloop/pisp/pull/55

**Comment by lewisdaly** [comment]
Closing this in favour of: https://github.com/mojaloop/pisp/issues/71
